### PR TITLE
[YUNIKORN-1575] [core] Ability to specify guaranteedResources via namespace annotations

### PR DESCRIPTION
**Pull request by zhuqi-lucas**
### What is this PR for?
The core side change to support ability to specify guaranteedResources via namespace annotations.

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
* Open an issue on Jira https://issues.apache.org/jira/browse/YUNIKORN-1575
* Put link here, and add [YUNIKORN-*Jira number*] in PR title, eg. `[YUNIKORN-2] Gang scheduling interface parameters`

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Review by wilfred-s — CHANGES_REQUESTED** [review]
We need to add a test for the guaranteed tag to return an empty string (693) and if all resources are 0 (703) we have those for the max tag should be doable for the guaranteed tag too.

File: pkg/scheduler/objects/queue.go [hunk 1]
```diff
@@ -663,29 +663,61 @@ func (sq *Queue) AddApplication(app *Application) {
...
-		return
+	if quota != "" {
+		quotaRes, quotaErr = resources.NewResourceFromString(quota)
+		if quotaErr != nil {
+			log.Log(log.SchedQueue).Warn("application resource quota conversion failure",
+				zap.String("json quota string", quota),
+				zap.Error(quotaErr))
+		} else if !resources.StrictlyGreaterThanZero(quotaRes) {
+			log.Log(log.SchedQueue).Warn("application resource quota has at least one 0 value: cannot set queue limit",
+				zap.Stringer("maxResource", quotaRes))
```
Comment (manirajv06) [inline pkg/scheduler/objects/queue.go:684]: Error message bit misleading. Many be "Max resource quantities should be greater than zero" ?

File: pkg/scheduler/objects/queue.go [hunk 2]
```diff
@@ -663,29 +663,61 @@ func (sq *Queue) AddApplication(app *Application) {
 	// YUNIKORN-199: update the quota from the namespace
 	// get the tag with the quota
 	quota := app.GetTag(common.AppTagNamespaceResourceQuota)
-	if quota == "" {
+	// get the tag with the guaranteed resource
+	guaranteed := app.GetTag(common.AppTagNamespaceResourceGuaranteed)
+	if quota == "" && guaranteed == "" {
```
Comment (manirajv06) [inline pkg/scheduler/objects/queue.go:668]: Is guaranteed mandatory? or Use only when annotation has it?

**Review by manirajv06 — CHANGES_REQUESTED** [review]
Overall looks good. Given some minor comments

File: pkg/scheduler/objects/queue.go [hunk 3]
```diff
@@ -663,29 +663,61 @@ func (sq *Queue) AddApplication(app *Application) {
...
+				zap.String("json guaranteed string", guaranteed),
+				zap.Error(guaranteedErr))
+		} else if !resources.StrictlyGreaterThanZero(guaranteedRes) {
+			log.Log(log.SchedQueue).Warn("application guaranteed resource has at least one 0 value: cannot set queue guaranteed resource",
+				zap.Stringer("guaranteedResource", guaranteedRes))
+			guaranteedRes = nil // Skip setting guaranteed resource if it has a value <= 0
+		}
+	}
+
+	if quotaErr != nil && guaranteedErr != nil {
```
[the remark this answers is not in the document]
Comment (zhuqi-lucas) [reply pkg/scheduler/objects/queue.go:703]: I think we will return when both quotaErr and guaranteedErr, for example:
When quotaErr happen but guaranteedErr not happen, we can set for the guranteed resource seperately, every annotaion can set independently

So we only return here when both quotaErr and guaranteedErr happend.

[the remark this answers is not in the document]
Comment (zhuqi-lucas) [reply pkg/scheduler/objects/queue.go:703]: Sure, will address the return when logging.

File: pkg/scheduler/objects/queue.go (hunk 1 above)
Comment (zhuqi-lucas) [reply pkg/scheduler/objects/queue.go:684]: Good suggestion, will address it.

File: pkg/scheduler/objects/queue.go (hunk 2 above)
Comment (zhuqi-lucas) [reply pkg/scheduler/objects/queue.go:668]: No @manirajv06 , it only uses when user annotation it just like quota, every annotation can be used independently.

Comment (manirajv06) [reply pkg/scheduler/objects/queue.go:668]: If we can treat each annotation independently, then we should not put together in a single if checks. We can have separate if checks for each one.

File: pkg/scheduler/objects/queue.go [hunk 4]
```diff
@@ -663,29 +663,60 @@ func (sq *Queue) AddApplication(app *Application) {
...
-		return
+
+	// need to set guaranteed resource: convert json string to resource
+	if guaranteed != "" {
+		guaranteedRes, guaranteedErr = resources.NewResourceFromString(guaranteed)
+		if guaranteedErr != nil {
+			log.Log(log.SchedQueue).Warn("application guaranteed resource conversion failure",
+				zap.String("json guaranteed string", guaranteed),
+				zap.Error(guaranteedErr))
+			if quotaErr != nil {
```
Comment (manirajv06) [inline pkg/scheduler/objects/queue.go:696]: guaranteedErr, not quotaErr :)

**Review by manirajv06 — CHANGES_REQUESTED** [review]

File: pkg/scheduler/objects/queue.go (hunk 4 above)
Comment (zhuqi-lucas) [reply pkg/scheduler/objects/queue.go:696]: I think we should use quotaErr here, because we only need to return for both quotaErr != nil and guaranteedErr != nil

**Review by manirajv06 — APPROVED** [review]

**Comment by manirajv06** [comment]
looks good. See if you can increase the test coverage.

**Comment by zhuqi-lucas** [comment]
Thank you @manirajv06 for review, i wanted to increase the test coverage here, but it is the log error lines, how can we easily cover the log in unit test?
